File: src/rust-media-engine/src/media/video.rs
use crate::media::ConversionResult;
use std::path::Path;
use tokio::fs;
use url::Url;

const DASH_NAMESPACE: &str = "urn:mpeg:dash:schema:mpd:2011";
const DASH_TIMEOUT_SECS: u64 = 15;

/// Resolved video and audio stream URLs from a v.redd.it video.
pub struct DashStreams {
    pub video_url: Option<String>,
    pub audio_url: Option<String>,
}

/// Get the best video and audio streams for a v.redd.it URL.
/// Mirrors Python VideoHandler._get_best_streams exactly:
///   1. Try DASHPlaylist.mpd first and parse it
///   2. Fall back to probing direct quality URLs + audio discovery
pub async fn get_best_streams(
    client: &reqwest::Client,
    video_url: &str,
) -> DashStreams {
    // Only handle v.redd.it base URLs (not direct .mp4/.mpd files)
    if video_url.contains("v.redd.it")
        && !video_url.ends_with(".mp4")
        && !video_url.ends_with(".mpd")
    {
        let base_url = video_url.trim_end_matches('/');
        let dash_url = format!("{}/DASHPlaylist.mpd", base_url);

        // Try DASH manifest first (like Python)
        if url_exists(client, &dash_url).await {
            log::debug!("Found DASH manifest: {}", dash_url);
            if let Some(streams) = parse_dash_playlist(client, &dash_url).await {
                return streams;
            }
        }

        // Fall back to direct quality URL probing (like Python)
        let qualities = ["DASH_1080.mp4", "DASH_720.mp4", "DASH_480.mp4"];
        for quality in qualities {
            let test_url = format!("{}/{}", base_url, quality);
            if url_exists(client, &test_url).await {
                let audio_url = find_audio_stream(client, &test_url).await;
                return DashStreams {
                    video_url: Some(test_url),
                    audio_url,
                };
            }
        }

        return DashStreams {
            video_url: None,
            audio_url: None,
        };
    }

    // If URL ends with .mpd or contains DASHPlaylist, parse it directly
    if video_url.ends_with(".mpd") || video_url.contains("DASHPlaylist") {
        if let Some(streams) = parse_dash_playlist(client, video_url).await {
            return streams;
        }
    }

    // Direct video URL, no audio
    DashStreams {
        video_url: Some(video_url.to_string()),
        audio_url: None,
    }
}

/// Parse a DASH manifest and extract best video and audio stream URLs.
/// Mirrors Python VideoHandler._parse_dash_playlist.
async fn parse_dash_playlist(
    client: &reqwest::Client,
    dash_url: &str,
) -> Option<DashStreams> {
    let response = client
        .get(dash_url)
        .timeout(std::time::Duration::from_secs(DASH_TIMEOUT_SECS))
        .send()
        .await
        .ok()?;

    if !response.status().is_success() {
        return None;
    }

    let content = response.text().await.ok()?;
    let doc = roxmltree::Document::parse(&content).ok()?;

    let base_url = get_base_url(dash_url);

    let mut video_streams: Vec<(u32, String)> = Vec::new();
    let mut audio_streams: Vec<String> = Vec::new();

    // Find all AdaptationSet elements (with namespace)
    for node in doc.descendants() {
        if node.tag_name().name() == "AdaptationSet"
            && node.tag_name().namespace() == Some(DASH_NAMESPACE)
        {
            let content_type = node
                .attribute("contentType")
                .unwrap_or("")
                .to_lowercase();

            // Find Representation elements within this AdaptationSet
            for child in node.descendants() {
                if child.tag_name().name() == "Representation"
                    && child.tag_name().namespace() == Some(DASH_NAMESPACE)
                {
                    if let Some(media_url) = extract_media_url(&child, &base_url) {
                        if content_type.contains("video") {
                            let height = child
                                .attribute("height")
                                .and_then(|h| h.parse::<u32>().ok())
                                .unwrap_or(480);
                            video_streams.push((height, media_url));
                        } else if content_type.contains("audio") {
                            audio_streams.push(media_url);
                        }
                    }
                }
            }
        }
    }

    // Pick best video (highest resolution) and first audio
    let best_video = video_streams
        .iter()
        .max_by_key(|(height, _)| *height)
        .map(|(_, url)| url.clone());

    let best_audio = audio_streams.into_iter().next();

    log::debug!(
        "DASH parsed - video: {:?}, audio: {:?}",
        best_video,
        best_audio
    );

    Some(DashStreams {
        video_url: best_video,
        audio_url: best_audio,
    })
}

/// Extract media URL from a DASH Representation element.
/// Mirrors Python VideoHandler._extract_media_url.
fn extract_media_url(representation: &roxmltree::Node, base_url: &str) -> Option<String> {
    // First try BaseURL child element
    for child in representation.children() {
        if child.tag_name().name() == "BaseURL" {
            if let Some(text) = child.text() {
                let trimmed = text.trim();
                if !trimmed.is_empty() {
                    return join_url(base_url, trimmed);
                }
            }
        }
    }

    // Fall back to constructing from representation attributes
    let rep_id = representation.attribute("id").unwrap_or("");
    let height = representation.attribute("height");

    if rep_id.to_lowercase().contains("audio") {
        return join_url(base_url, "DASH_audio.mp4");
    } else if let Some(h) = height {
        return join_url(base_url, &format!("DASH_{}.mp4", h));
    }

    None
}

/// Get base URL from a DASH manifest URL (everything up to the last path segment).
/// Mirrors Python VideoHandler._get_base_url.
fn get_base_url(dash_url: &str) -> String {
    if let Ok(parsed) = Url::parse(dash_url) {
        let mut path_segments: Vec<&str> = parsed.path().split('/').collect();
        // Remove the last segment (the filename)
        if path_segments.len() > 1 {
            path_segments.pop();
        }
        let base_path = path_segments.join("/");
        format!("{}://{}{}/", parsed.scheme(), parsed.host_str().unwrap_or(""), base_path)
    } else {
        // Fallback: strip everything after the last /
        match dash_url.rfind('/') {
            Some(pos) => format!("{}/", &dash_url[..pos]),
            None => dash_url.to_string(),
        }
    }
}

/// Join a base URL with a relative path.
fn join_url(base: &str, relative: &str) -> Option<String> {
    if let Ok(base_parsed) = Url::parse(base) {
        base_parsed.join(relative).ok().map(|u| u.to_string())
    } else {
        // Simple fallback
        Some(format!("{}{}", base.trim_end_matches('/'), if relative.starts_with('/') { "" } else { "/" }).to_string()
            + relative)
    }
}

/// Find audio stream URL by probing known audio filenames.
/// Mirrors Python VideoHandler._find_audio_stream.
async fn find_audio_stream(client: &reqwest::Client, video_url: &str) -> Option<String> {
    if let Ok(parsed) = Url::parse(video_url) {
        let path = parsed.path();
        if path.contains("DASH_") {
            let base_path = path.split("DASH_").next()?;
            let base_url = format!(
                "{}://{}{}",
                parsed.scheme(),
                parsed.host_str().unwrap_or(""),
                base_path
            );

            let audio_formats = [
                "DASH_audio.mp4",
                "DASH_AUDIO_128.mp4",
                "DASH_128.mp4",
                "audio",
            ];

            for audio_format in audio_formats {
                let audio_url = format!("{}{}", base_url, audio_format);
                if url_exists(client, &audio_url).await {
                    log::debug!("Found audio stream: {}", audio_url);
                    return Some(audio_url);
                }
            }
        }
    }
    None
}

/// Check if a URL exists (HEAD request returns 200).
async fn url_exists(client: &reqwest::Client, url: &str) -> bool {
    match client
        .head(url)
        .timeout(std::time::Duration::from_secs(10))
        .send()
        .await
    {
        Ok(resp) => resp.status().is_success(),
        Err(_) => false,
    }
}

/// Process video+audio streams with FFmpeg.
/// FFmpeg command matches Python VideoHandler._merge_sync / _convert_sync exactly.
pub async fn process_video_streams(
    video_url: String,
    audio_url: Option<String>,
    output_path: String,
) -> Result<ConversionResult, Box<dyn std::error::Error + Send + Sync>> {
    // Ensure output directory exists
    let path = Path::new(&output_path);
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent).await?;
    }

    let mut cmd = tokio::process::Command::new("ffmpeg");

    if let Some(audio) = &audio_url {
        // Exact same command as Python _merge_sync
        cmd.args([
            "-i",
            &video_url,
            "-i",
            audio,
            "-c:v",
            "libvpx-vp9",
            "-crf",
            "23",
            "-b:v",
            "0",
            "-row-mt",
            "1",
            "-cpu-used",
            "2",
            "-c:a",
            "libopus",
            "-b:a",
            "128k",
            "-shortest",
            "-threads",
            "0",
            "-y",
            &output_path,
        ]);
    } else {
        // Exact same command as Python _convert_sync
        cmd.args([
            "-i",
            &video_url,
            "-c:v",
            "libvpx-vp9",
            "-crf",
            "23",
            "-b:v",
            "0",
            "-row-mt",
            "1",
            "-cpu-used",
            "2",
            "-an",
            "-threads",
            "0",
            "-y",
            &output_path,
        ]);
    }

    let output = cmd.output().await?;

    if output.status.success() {
        let converted_size = fs::metadata(&output_path).await?.len();

        Ok(ConversionResult {
            success: true,
            output_path: Some(output_path),
            original_size: 0,
            converted_size,
            error: None,
        })
    } else {
        let error = String::from_utf8_lossy(&output.stderr);
        Ok(ConversionResult {
            success: false,
            output_path: None,
            original_size: 0,
            converted_size: 0,
            error: Some(format!("Video processing failed: {}", error)),
        })
    }
}

File: src/reddit_scraper/media/video.py
"""
Video handler - async version for Reddit videos with quality selection.
"""

import asyncio
import subprocess
import xml.etree.ElementTree as ET
from pathlib import Path
from typing import List, Optional, Tuple
from urllib.parse import urljoin, urlparse

from .base import BaseMediaHandler, MediaConfig, ProcessingResult


class VideoHandler(BaseMediaHandler):
    """Handle Reddit video downloads with quality selection (async)."""

    def __init__(self, subreddit: str):
        super().__init__(subreddit, "video")

    async def download_and_process(self, video_url: str, item_id: str) -> ProcessingResult:
        """Download and process Reddit video."""
        try:
            output_path = self.media_dir / f"video_{item_id}.webm"

            if await self.check_file_exists(output_path):
                return ProcessingResult(
                    url=video_url,
                    item_id=item_id,
                    success=True,
                    path=str(output_path),
                    skipped=True,
                )

            self.logger.debug(f"Processing video: {video_url}")

            # Get best quality streams
            video_stream_url, audio_stream_url = await self._get_best_streams(video_url)

            if not video_stream_url:
                return ProcessingResult(
                    url=video_url, item_id=item_id, success=False, error="No video stream found"
                )

            # Process based on available streams
            if audio_stream_url:
                self.logger.debug("Found video + audio streams, merging...")
                result_path = await self._merge_video_audio(
                    video_stream_url, audio_stream_url, output_path
                )
            else:
                self.logger.debug("Video-only stream, converting...")
                result_path = await self._convert_video_only(video_stream_url, output_path)

            if result_path:
                return ProcessingResult(
                    url=video_url, item_id=item_id, success=True, path=result_path
                )
            else:
                return ProcessingResult(
                    url=video_url, item_id=item_id, success=False, error="FFmpeg processing failed"
                )

        except Exception as e:
            self.logger.error(f"Video processing failed for {video_url}: {e}")
            return ProcessingResult(url=video_url, item_id=item_id, success=False, error=str(e))

    async def _get_best_streams(self, video_url: str) -> Tuple[Optional[str], Optional[str]]:
        """Extract best quality video and audio streams."""
        try:
            if "v.redd.it" in video_url and not video_url.endswith((".mp4", ".mpd")):
                base_url = video_url.rstrip("/")
                dash_url = f"{base_url}/DASHPlaylist.mpd"

                if await self._url_exists(dash_url):
                    self.logger.debug(f"Found DASH manifest: {dash_url}")
                    return await self._parse_dash_playlist(dash_url)

                # Try direct quality URLs
                for quality in ["DASH_1080.mp4", "DASH_720.mp4", "DASH_480.mp4"]:
                    test_url = f"{base_url}/{quality}"
                    if await self._url_exists(test_url):
                        audio_url = await self._find_audio_stream(test_url)
                        return test_url, audio_url

                return None, None

            elif video_url.endswith(".mpd") or "DASHPlaylist" in video_url:
                return await self._parse_dash_playlist(video_url)

            else:
                return video_url, None

        except Exception as e:
            self.logger.error(f"Stream extraction error: {e}")
            return video_url, None

    async def _parse_dash_playlist(self, dash_url: str) -> Tuple[Optional[str], Optional[str]]:
        """Parse DASH manifest for best streams."""
        try:
            response = await asyncio.to_thread(
                self.session.get, dash_url, timeout=MediaConfig.DASH_TIMEOUT
            )
            response.raise_for_status()
            content = response.content

            root = ET.fromstring(content)
            base_url = self._get_base_url(dash_url)

            video_streams = []
            audio_streams = []

            adaptation_sets = root.findall(".//{urn:mpeg:dash:schema:mpd:2011}AdaptationSet")

            for adaptation_set in adaptation_sets:
                content_type = adaptation_set.get("contentType", "").lower()
                representations = adaptation_set.findall(
                    ".//{urn:mpeg:dash:schema:mpd:2011}Representation"
                )

                for representation in representations:
                    media_url = self._extract_media_url(representation, base_url)

                    if media_url:
                        if "video" in content_type:
                            height = representation.get("height")
                            quality = int(height) if height else 480
                            video_streams.append((quality, media_url))
                        elif "audio" in content_type:
                            audio_streams.append(media_url)

            best_video = max(video_streams, key=lambda x: x[0])[1] if video_streams else None
            best_audio = audio_streams[0] if audio_streams else None

            self.logger.debug(f"Selected video: {best_video}, audio: {best_audio}")
            return best_video, best_audio

        except Exception as e:
            self.logger.error(f"DASH parsing failed: {e}")
            return None, None

    def _get_base_url(self, dash_url: str) -> str:
        """Get base URL for relative paths."""
        parsed = urlparse(dash_url)
        return f"{parsed.scheme}://{parsed.netloc}{'/'.join(parsed.path.split('/')[:-1])}/"

    def _extract_media_url(self, representation, base_url: str) -> Optional[str]:
        """Extract media URL from DASH representation."""
        try:
            base_url_elem = representation.find(".//{urn:mpeg:dash:schema:mpd:2011}BaseURL")
            if base_url_elem is not None and base_url_elem.text:
                return urljoin(base_url, base_url_elem.text.strip())

            representation_id = representation.get("id", "")
            height = representation.get("height")

            if "audio" in representation_id.lower():
                return urljoin(base_url, "DASH_audio.mp4")
            elif height:
                return urljoin(base_url, f"DASH_{height}.mp4")

            return None

        except Exception:
            return None

    async def _find_audio_stream(self, video_url: str) -> Optional[str]:
        """Find corresponding audio stream."""
        try:
            parsed = urlparse(video_url)

            if "DASH_" in parsed.path:
                base_path = parsed.path.split("DASH_")[0]
                base_url = f"{parsed.scheme}://{parsed.netloc}{base_path}"

                for audio_format in [
                    "DASH_audio.mp4",
                    "DASH_AUDIO_128.mp4",
                    "DASH_128.mp4",
                    "audio",
                ]:
                    audio_url = f"{base_url}{audio_format}"
                    if await self._url_exists(audio_url):
                        self.logger.debug(f"Found audio stream: {audio_url}")
                        return audio_url

            return None

        except Exception:
            return None

    async def _url_exists(self, url: str) -> bool:
        """Check if URL exists."""
        try:
            response = await asyncio.to_thread(
                self.session.head, url, timeout=MediaConfig.HEAD_REQUEST_TIMEOUT
            )
            return response.status_code == 200
        except Exception:
            return False

    async def _merge_video_audio(
        self, video_url: str, audio_url: str, output_path: Path
    ) -> Optional[str]:
        """Merge video and audio streams (runs in thread)."""
        return await asyncio.to_thread(self._merge_sync, video_url, audio_url, output_path)

    def _merge_sync(self, video_url: str, audio_url: str, output_path: Path) -> Optional[str]:
        """Synchronous FFmpeg merge."""
        try:
            cmd = [
                "ffmpeg",
                "-i",
                video_url,
                "-i",
                audio_url,
                "-c:v",
                "libvpx-vp9",
                "-crf",
                str(MediaConfig.VIDEO_CRF),
                "-b:v",
                "0",
                "-row-mt",
                "1",
                "-cpu-used",
                str(MediaConfig.VIDEO_CPU_USED),
                "-c:a",
                "libopus",
                "-b:a",
                MediaConfig.AUDIO_BITRATE,
                "-shortest",
                "-threads",
                "0",
                "-y",
                str(output_path),
            ]

            result = subprocess.run(cmd, capture_output=True, text=True, timeout=600)

            if result.returncode == 0 and output_path.exists():
                file_size = output_path.stat().st_size
                self.logger.info(
                    f"Video+Audio merge complete: {output_path.name} ({file_size:,} bytes)"
                )
                return str(output_path)
            else:
                self.logger.error(f"Merge failed: {result.stderr}")
                return None

        except subprocess.TimeoutExpired:
            self.logger.error(f"Video merge timeout for {output_path.name}")
            return None
        except Exception as e:
            self.logger.error(f"Merge error: {e}")
            return None

    async def _convert_video_only(self, video_url: str, output_path: Path) -> Optional[str]:
        """Convert video-only stream (runs in thread)."""
        return await asyncio.to_thread(self._convert_sync, video_url, output_path)

    def _convert_sync(self, video_url: str, output_path: Path) -> Optional[str]:
        """Synchronous FFmpeg conversion."""
        try:
            cmd = [
                "ffmpeg",
                "-i",
                video_url,
                "-c:v",
                "libvpx-vp9",
                "-crf",
                str(MediaConfig.VIDEO_CRF),
                "-b:v",
                "0",
                "-row-mt",
                "1",
                "-cpu-used",
                str(MediaConfig.VIDEO_CPU_USED),
                "-an",
                "-threads",
                "0",
                "-y",
                str(output_path),
            ]

            result = subprocess.run(cmd, capture_output=True, text=True, timeout=300)

            if result.returncode == 0 and output_path.exists():
                file_size = output_path.stat().st_size
                self.logger.info(
                    f"Video conversion complete: {output_path.name} ({file_size:,} bytes)"
                )
                return str(output_path)
            else:
                self.logger.error(f"Conversion failed: {result.stderr}")
                return None

        except subprocess.TimeoutExpired:
            self.logger.error(f"Video conversion timeout for {output_path.name}")
            return None
        except Exception as e:
            self.logger.error(f"Conversion error: {e}")
            return None

    async def batch_process(
        self, url_id_pairs: List[tuple], max_concurrent: Optional[int] = None
    ) -> dict:
        """Process multiple videos concurrently."""
        if max_concurrent is None:
            max_concurrent = MediaConfig.MAX_CONCURRENT_VIDEOS

        result = await self.batch_process_with_progress(
            url_id_pairs, self.download_and_process, max_concurrent
        )

        return result.to_dict()
